plugins/inputs/directory_monitor/directory_monitor.go

```diff
@@ -0,0 +1,468 @@
+			timeThresholdExceeded := time.Since(stat.AccessTime()) > monitor.DirectoryDurationThreshold.Duration
+			_, fileAlreadyProcessing := monitor.filesInUse.Get(filePath)
+
+			// If file is decaying, process it.
+			if enoughRoomForFile && timeThresholdExceeded && !fileAlreadyProcessing {
+				// Set the file as 'in use' so that subsequent Monitor runs won't possibly pick it up again.
+				monitor.filesInUse.Set(filePath, struct{}{})
+				filesToProcess = append(filesToProcess, fileInfo)
+			}
+		}
+
+		monitor.processFileBatch(filesToProcess, acc)
+
+		select {
+		// Monitor in intervals.
+		case <-time.After(monitor.MonitorInterval.Duration):
+		// Allow the monitor to be quit.
+		case <-monitor.quit:
+			return
+		}
+	}
+}
+
+func (monitor *DirectoryMonitor) processFileBatch(files []os.FileInfo, acc telegraf.Accumulator) {
+	// Process each valid file with a new goroutine.
+	for _, fileInfo := range files {
+		if fileInfo.IsDir() {
+			continue
+		}
+
+		filePath := monitor.Directory + "/" + fileInfo.Name()
+
+		// File must be configured to be monitored, if any configuration...
+		if !monitor.isMonitoredFile(fileInfo.Name()) {
+			continue
+		}
+
+		// ...and should not be configured to be ignored.
+		if monitor.isIgnoredFile(fileInfo.Name()) {
+			continue
+		}
+
+		monitor.waitGroup.Add(1)
+		go monitor.read(filePath, monitor.waitGroup)
+	}
+}
+
+func (monitor *DirectoryMonitor) read(filePath string, waitGroup *sync.WaitGroup) {
+	// Remove the file from the set of files in use when it's finished.
+	defer monitor.filesInUse.Remove(filePath)
+	defer waitGroup.Done()
+
+	// Open, read, and parse the contents of the file.
+	err := monitor.ingestFile(filePath)
+
+	// Handle a file read error. We don't halt execution but do document, log, and move the problematic file.
+	if err != nil {
+		monitor.Log.Errorf("Error while reading file: '" + filePath + "'. " + err.Error())
+		monitor.filesDropped.Incr(1)
+		if monitor.UseErrorDirectory {
+			monitor.moveFile(filePath, monitor.ErrorDirectory)
+		}
+		return
+	}
+
+	// File is finished, move it to the 'finished' directory.
+	monitor.moveFile(filePath, monitor.FinishedDirectory)
+	monitor.filesProcessed.Incr(1)
+}
+
+func (monitor *DirectoryMonitor) ingestFile(filePath string) error {
+	file, err := os.Open(filePath)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	parser, err := monitor.parserFunc()
+	if err != nil {
+		return fmt.Errorf("E! Creating parser: %s", err.Error())
+	}
+
+	// Handle gzipped files.
+	var reader io.Reader
+	if filepath.Ext(filePath) == ".gz" {
+		reader, err = gzip.NewReader(file)
+		if err != nil {
+			return err
+		}
+	} else {
+		reader, _ = utfbom.Skip(monitor.decoder.Reader(file))
+	}
+
+	return monitor.parseFile(parser, reader)
+}
+
+func (monitor *DirectoryMonitor) parseFile(parser parsers.Parser, reader io.Reader) error {
+	// JSON files can't be read line-by-line with the current parser.
```

Comment:
check other plugins. this is supported elsewhere. Reading the whole file into memory for a data dump is not an option. needs further discussion.